yarnitor/background/worker.py

```diff
@@ -34,260 +38,390 @@
+atexit.register(lambda: threadpool.shutdown(False))
 
 class Progress(object):
-    """Utility class for storing progress information
-    """
-
+    """Utility class for storing mutable progress information."""
     def __init__(self, name, completed=0, failed=0, running=0, total=0):
         self.name = name
         self.completed = completed
         self.failed = failed
         self.running = running
         self.total = total
 
+    def to_dict(self):
+        return self.__dict__
 
-class YarnApi(object):
-    """Collection of yarn endpoints we care about.
-    """
 
+class YARNHandler(object):
+    """Manages HTTP communication with a YARN ResourceManager (RM) to fetch
+    information about applications.
+
+    Parameters
+    ----------
+    host: str
+        Protocol, hostname, and optional port of the YARN RM
+    version: str, optional
+        Version of the YARN RM API
+
+    Attributes
+    ----------
+    base_url: dict
+        'host' and 'version' of the YARN RM
+    """
     def __init__(self, host, version="v1"):
-        self.host = host
-        self.version = version
+        self.base_url = {
+            'host': host,
+            'version': version
+        }
+
+    def get_url(self, path, **params):
+        """Issues an HTTP GET to the given path with the given parameters
+        and treats the response as JSON.
 
-    def get_url(self, url, **params):
-        final_url = "{host}/ws/{version}".format(**self.__dict__) + url
-        resp = requests.get(final_url, params)
+        Parameters
+        ----------
+        path: str
+            Path to append to the root YARN RM path
+        **params, dict
+            Query parameters to append to the YARN RM URL
+
+        Returns
+        -------
+        dict
+            JSON decoded response
+        """
+        # Handle HA redirects in a loop so that we can control how many
+        # we're willing to follow and don't blow out the stack with recursion.
+        for i in range(MAX_HA_REDIRECTS):
+            final_url = "{host}/ws/{version}/{path}".format(path=path, **self.base_url)
+            resp = requests.get(final_url, params)
+            resp.raise_for_status()
+            # YARN uses the Refresh header to indicate a change in the primary RM
+            ha_redirect = resp.headers.get('Refresh')
+            if ha_redirect is None:
+                break
+            else:
+                # Store the new host in the instance for the next request
+                # It comes in the form "3; url=http://newhost:port/same/path/as/request?args"
+                _, key = ha_redirect.split(';')
+                _, new_url = key.strip().split('url=')
+                parsed = urlparse(new_url.strip())
+                new_host = '{parsed.scheme}://{parsed.netloc}'.format(parsed=parsed)
+                self.base_url['host'] = new_host
+                logger.warn('YARN HA redirect, switching to URL: %s', new_host)
```

Comment:
This doesn't feel like a warn. Maybe info, but perhaps even debug?

Reply:
Maybe. It does mean the primary ResourceManager changed from what was configured.